django/utils/hashable.py

```diff
@@ -1,9 +1,16 @@
+from django.utils.itercompat import is_iterable
+
+
 def make_hashable(value):
-    if isinstance(value, list):
-        return tuple(map(make_hashable, value))
     if isinstance(value, dict):
         return tuple([
             (key, make_hashable(nested_value))
             for key, nested_value in value.items()
         ])
+    try:
+        hash(value)
+    except TypeError:
+        if is_iterable(value):
+            return tuple(map(make_hashable, value))
+        raise
```

Comment:
Thanks, I'll add some comments based on what you wrote.

Just need something to test the raise branch. It looks like Python modules are hashable unless I've misunderstood what you meant:
```
>>> import django
>>> django
<module 'django' from '/home/tim/code/django/django/__init__.py'>
>>> hash(django)
-9223363272599355027
```

Reply:
Yep! Just tried to hash dozens of python modules and they're all hashable! I think since object implements `__hash__`, most modules are going to be hashable? Immutable collection types would throw the error, but since we solve those cases specifically, I'm not sure what else will throw the error. 

List for example has hash defined like this:
```
    __hash__ = None
```

Reply:
Oh, I didn't know that! You could define an explicitly unhashable type

```python
class Unhashable:
    def __hash__(self):
        raise TypeError('Unhashable')
```

Reply:
We could simplify that to be similar to how the builtin defines `__hash__`. I'll commit that up now!

Reply:
Actually `set` isn't hashable but your point remains for `frozenset`. Commit two makes sets hashable by converting them to tuple. I added tests for those cases.

I added `test_bad_hash_typerror_not_hidden` to hit the `raise`. If you have something better to suggest, I'll change it.

Reply:
Oops, we're racing -- I just pushed some changes. Feel free to "simplify that to be similar to how the builtin defines `__hash__`" as I'm not sure what you mean."

Reply:
Looks like you can set `__hash__ = None` to make objects unhashable. I think that's what @aspalding had in mind.

Reply:
@charettes exactly. So we could define our test case like:
```
        class Test:
            __hash__ = None
```

 I can commit that up, or we can just go with the current implementation. Up to you @timgraham.

Reply:
Done, I think it's good to go now.